In [1]:
import pandas as pd
import numpy as np
import os 
import warnings
warnings.filterwarnings("ignore")

In [104]:
df = pd.read_csv("../Data/flight_price.csv")

In [3]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [5]:
# OBSERVATION :-
# The data set contains 10,683 rows and eleven features.
# It also have single null value in Route and Total_Stops.
# The data type of few columns are not correct:-
        # The data type of columns such as Date_of_Journey, Dep_Time, Arrival_Time should be DATETIME but it is currently OBJECT.
        # The data type of columns such as Total_Stops, Duration should be INTEGER but it is currently OBJECT.

## 3. PRELIMINARY ANALYSIS

### 3.1 CHECKING DATATYPE

In [6]:
df.head(2)

Airline Date_of_Journey    Source Destination                  Route  \
0     IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1  Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662

In [7]:
df.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [8]:
def maxMinLength(df, column):
    maxi = 0
    mini = 100 
    for i in df[column]:

        maxi = max(maxi, len(i))
        mini = min(mini, len(i))

    print("MAXIMUM LENGTH :-", maxi)
    print("MINIMUM LENGTH :-", mini)
    return maxi, mini

def toFindUniqueLengthValueCount(df, column):
    Dictionary = {}  

    for i in df[column]:  
        length = len(str(i))  

        if length in Dictionary:
            Dictionary[length] += 1  
        else:
            Dictionary[length] = 1  

    return Dictionary

def toPrintUniqueValue(df, columns):
    mySet = set()
    for i in df[columns]:

        try:
            if len(i) not in mySet:
                print(len(i),"->",  i)
                mySet.add(len(i))
        except:
            print(i)

##### 3.1.1. "Date_of_Journey" columns

In [9]:
df['Date_of_Journey'].iloc[0]

'24/03/2019'

In [10]:
df['Date_of_Journey'].iloc[1]

'1/05/2019'

In [11]:
maxi, mini = maxMinLength(df, 'Date_of_Journey')

MAXIMUM LENGTH :- 10
MINIMUM LENGTH :- 9


In [12]:
# OBSERVATION:- 
#     FROM THIS WE OBSERVE THAT THERE IS NO EXTRA DATA PRESENT IN DATA_OF_JOURNEY
#     INSTANCES IT JUST CONSISTS OF DATA ABOUT EXACT DATE 
#     IF DATE < 10: LENGTH == 9
#     ELSE : LENGTH == 10    
#     Thus, there is no need to clean we just need to convert datatype

##### 3.1.2. "Dep_Time" columns

In [13]:
df['Dep_Time'].iloc[0]

'22:20'

In [14]:
df['Dep_Time'].iloc[100]

'08:25'

In [15]:
maxi, mini = maxMinLength(df, 'Dep_Time')

MAXIMUM LENGTH :- 5
MINIMUM LENGTH :- 5


In [16]:
# OBSERVATION:- 
#     As the length remains same and it just consists of require imformation. 
#     Thus, there is no need to clean we just need to convert datatype

##### 3.1.3. "Arrival_Time" columns

In [17]:
string = df['Arrival_Time'].iloc[0]
print("string :- ", string)
print("length : ", len(string))

string :-  01:10 22 Mar
length :  12


In [18]:
string = df['Arrival_Time'].iloc[1]
print("string :- ", string)
print("length : ", len(string))

string :-  13:15
length :  5


In [19]:
maxi, mini = maxMinLength(df, 'Arrival_Time')

MAXIMUM LENGTH :- 12
MINIMUM LENGTH :- 5


In [20]:
print("Unique Length are:-", toFindUniqueLengthValueCount(df, 'Arrival_Time'))        

Unique Length are:- {12: 4335, 5: 6348}


In [21]:
# OBSERVATION:- 
#     As the string contain some unnecessary information there is need to clean.
#     Thus, there ther is need to clean and convert datatype.

##### 3.1.4. "Duration" columns

In [22]:
df['Duration'].iloc[0]

'2h 50m'

In [23]:
df['Duration'].iloc[2000]

'8h 15m'

In [24]:
maxi, mini = maxMinLength(df, 'Duration')

MAXIMUM LENGTH :- 7
MINIMUM LENGTH :- 2


In [25]:
print("Unique Length are:-", toFindUniqueLengthValueCount(df, 'Duration'))        

Unique Length are:- {6: 5503, 3: 516, 7: 3884, 2: 516, 5: 264}


In [26]:
toPrintUniqueValue(df, 'Duration')

6 -> 2h 50m
3 -> 19h
7 -> 15h 30m
2 -> 3h
5 -> 3h 5m


In [27]:
# OBSERVATION:- 
#     All length are legit. 
#     The string contain some extra data but it is useful to change dataype  

##### 3.1.5. "Total_Stops" columns

In [28]:
string = df['Total_Stops'].iloc[0]
print("string :- ", string)
print("length : ", len(string))

string :-  non-stop
length :  8


In [29]:
print("Unique Length are:-", toFindUniqueLengthValueCount(df, 'Total_Stops'))        

Unique Length are:- {8: 3491, 7: 1566, 6: 5625, 3: 1}


In [30]:
toPrintUniqueValue(df, 'Total_Stops')

8 -> non-stop
7 -> 2 stops
6 -> 1 stop
nan


In [31]:
df['Total_Stops'].value_counts()

Total_Stops
1 stop      5625
non-stop    3491
2 stops     1520
3 stops       45
4 stops        1
Name: count, dtype: int64

In [32]:
# OBSERVATION:- 
#     As the string contain some unnecessary information there is need to clean.
#     Thus, there ther is need to clean and convert datatype and there is also 
#           a null value which we need to handle 

In [33]:
df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

##### 3.1.6. "Airline" columns

In [34]:
df['Airline'].value_counts()

Airline
Jet Airways                          3849
IndiGo                               2053
Air India                            1752
Multiple carriers                    1196
SpiceJet                              818
Vistara                               479
Air Asia                              319
GoAir                                 194
Multiple carriers Premium economy      13
Jet Airways Business                    6
Vistara Premium economy                 3
Trujet                                  1
Name: count, dtype: int64

##### 3.1.7. "Source" columns

In [35]:
df['Source'].value_counts()

Source
Delhi       4537
Kolkata     2871
Banglore    2197
Mumbai       697
Chennai      381
Name: count, dtype: int64

##### 3.1.8. "Destination" columns

In [36]:
df['Destination'].value_counts()

Destination
Cochin       4537
Banglore     2871
Delhi        1265
New Delhi     932
Hyderabad     697
Kolkata       381
Name: count, dtype: int64

##### 3.1.9. "Route" columns

In [37]:
df['Route'].value_counts()

Route
DEL → BOM → COK          2376
BLR → DEL                1552
CCU → BOM → BLR           979
CCU → BLR                 724
BOM → HYD                 621
                         ... 
CCU → VTZ → BLR             1
CCU → IXZ → MAA → BLR       1
BOM → COK → MAA → HYD       1
BOM → CCU → HYD             1
BOM → BBI → HYD             1
Name: count, Length: 128, dtype: int64

##### 3.1.10. "Additional_Info" columns

In [38]:
df['Additional_Info'].value_counts()

Additional_Info
No info                         8345
In-flight meal not included     1982
No check-in baggage included     320
1 Long layover                    19
Change airports                    7
Business class                     4
No Info                            3
1 Short layover                    1
Red-eye flight                     1
2 Long layover                     1
Name: count, dtype: int64

In [39]:
df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

- CONCLUSION :-
    - We just need to clean Date_of_Journey, Dep_Time, Arrival_Time, Duration 
    - and Total_Stops. rest all are in it actual form.

### 3.2 CHECK FOR DUPLICATES

In [40]:
df.duplicated().sum()

220

In [41]:
df.shape

(10683, 11)

In [42]:
df.loc[df.duplicated(keep=False)].sort_values(['Airline',
                                               'Date_of_Journey',
                                               'Source', 'Destination',
                                               'Dep_Time'
                                              ]).head(10)

Airline Date_of_Journey    Source Destination                  Route  \
6321   Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848   Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572    Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495   Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
9913   Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
3598   Air India       1/05/2019   Kolkata    Banglore  CCU → GAU → DEL → BLR   
10183  Air India       1/05/2019   Kolkata    Banglore  CCU → GAU → DEL → BLR   
4603   Air India       1/05/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
5042   Air India       1/05/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   

      Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
6321     08:50  23:55 02 Mar   39h 5m     2 stops         No info  17135  
9848     08:50  23:55 02 Mar   39h 5m     2 stops         No info  17135  
572      21:10         23:55   2h 45m    non-stop         No info   7591  
8168     21:10         23:55   2h 45m    non-stop         No info   7591  
1495     10:00  01:20 02 Apr  15h 20m     2 stops         No info  10408  
9913     10:00  01:20 02 Apr  15h 20m     2 stops         No info  10408  
3598     09:50  08:55 02 May   23h 5m     2 stops         No info  13227  
10183    09:50  08:55 02 May   23h 5m     2 stops         No info  13227  
4603     10:00  13:45 02 May  27h 45m     2 stops         No info  15164  
5042     10:00  13:45 02 May  27h 45m     2 stops         No info  15164

- From the above observation, we have confirmed that there are 220 duplicates,
- values which are present in our data set.

### 4. ANALYSING IN MORE DETAIL

In [43]:
[i for i in df['Airline'].unique()]

['IndiGo',
 'Air India',
 'Jet Airways',
 'SpiceJet',
 'Multiple carriers',
 'GoAir',
 'Vistara',
 'Air Asia',
 'Vistara Premium economy',
 'Jet Airways Business',
 'Multiple carriers Premium economy',
 'Trujet']

In [44]:
df['Source'].unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [45]:
[i for i in df['Additional_Info'].unique()]

['No info',
 'In-flight meal not included',
 'No check-in baggage included',
 '1 Short layover',
 'No Info',
 '1 Long layover',
 'Change airports',
 'Business class',
 'Red-eye flight',
 '2 Long layover']

- From this we can observe that just because of the case difference,
- the values are identifying uniquely.

### 5. CLEANING OPERATIONS

In [203]:
# METHOD CHAINING
# pandas method never affect actual data
# dictionary comprehension


def convertToMinutes(ser):
    return (
        ser
        .str.split(" ", expand=True) # expand turn series into dataframe
        .set_axis(["hour", "minute"], axis=1)
        .assign(
            hour=lambda df_: (
                df_
                .hour
                .str.replace("h", "")
                .astype(int)
                .mul(60)
            ),
            minute=lambda df_: (
                df_
                .minute
                .str.replace("m", "")
                .fillna("0")
                .astype(int)
            )
        )
        .sum(axis=1)
    )

def cleanData(df):
    extracted = df['Airline'].str.extract(r'^(.*?)\s*(Premium economy|Business)?$')
    return (
        df
        .drop(index=[6474])
        .drop_duplicates()
        .assign(**{
            col: df[col].str.strip()
            for col in df.select_dtypes(include='O').columns
        })
        .rename(columns=str.lower)
        .assign(
            airline=extracted[0].str.title(),  # Extracted Airline Name
            class_=extracted[1].fillna("Economy"),  # Extracted Class, filling NaN with "Economy"
            date_of_journey = lambda df_: pd.to_datetime(df.Date_of_Journey, dayfirst=True),
            dep_time = lambda df_: pd.to_datetime(df.Dep_Time).dt.time,
            arrival_time = lambda df_: pd.to_datetime(df.Arrival_Time.str.split(" ", n=1).str[0]).dt.time,
            duration = lambda df_: convertToMinutes(df_.duration),
            total_stops = lambda df_:(
                df_.total_stops
                .str.replace("non-stop", '0')
                .str.replace("2 stops", "2")
                .str.replace("1 stop", "1")
                .str.replace("3 stops", "3")
                .str.replace("4 stops", '4')
                .pipe(lambda ser: pd.to_numeric(ser))
            ),
            additional_info = lambda df_: df_.additional_info.replace("No info", "No Info")
        )
        .drop(columns='route')
    )



In [205]:
cleanData(df)

airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price   class_  
0         01:10:00       170          0.0         No Info   3897  Economy  
1         13:15:00       445          2.0         No Info   7662  Economy  
2         04:25:00      1140          2.0         No Info  13882  Economy  
3         23:30:00       325          1.0         No Info   6218  Economy  
4         21:35:00       285          1.0         No Info  13302  Economy  
...            ...       ...          ...             ...    ...      ...  
10678     22:25:00       150          0.0         No Info   4107  Economy  
10679     23:20:00       155          0.0         No Info   4145  Economy  
10680     11:20:00       180          0.0         No Info   7229  Economy  
10681     14:10:00       160          0.0         No Info  12648  Economy  
10682     19:15:00       500          2.0         No Info  11753  Economy  

[10462 rows x 11 columns]

In [206]:
cleanedData = cleanData(df)

In [207]:
cleanedData.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   airline          10462 non-null  object        
 1   date_of_journey  10462 non-null  datetime64[ns]
 2   source           10462 non-null  object        
 3   destination      10462 non-null  object        
 4   dep_time         10462 non-null  object        
 5   arrival_time     10462 non-null  object        
 6   duration         10462 non-null  int64         
 7   total_stops      10461 non-null  float64       
 8   additional_info  10462 non-null  object        
 9   price            10462 non-null  int64         
 10  class_           10462 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(7)
memory usage: 980.8+ KB


In [209]:
x = cleanedData.drop(columns='price')
y = cleanedData['price']

In [210]:
from sklearn.model_selection import train_test_split

In [211]:
X_, X_test, Y_, Y_test = train_test_split(x, y, test_size = 0.2, random_state=101)

In [212]:
X_train, X_val, Y_train, Y_val = train_test_split(X_, Y_, test_size=0.2, random_state=101)

In [213]:
print("The Shape of X_train:", X_train.shape)
print("The Shape of X_test:", X_test.shape)
print("The Shape of X_val:", X_val.shape)
print("The Shape of Y_test:", Y_test.shape)
print("The Shape of Y_train:", Y_train.shape)
print("The Shape of Y_val:", Y_val.shape)

The Shape of X_train: (6695, 10)
The Shape of X_test: (2093, 10)
The Shape of X_val: (1674, 10)
The Shape of Y_test: (2093,)
The Shape of Y_train: (6695,)
The Shape of Y_val: (1674,)


In [214]:
PROJECT_DIR = r"C:\Users\shrey\Desktop\DEEP LEARNING PROJECTS\3. flight Price Prediction"
DATA_DIR = "Data"

In [215]:
def export_data(X, Y, name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR, DATA_DIR, file_name)
    
    X.join(Y).to_csv(file_path, index = False)
    return pd.read_csv(file_path).head()

In [216]:
export_data(X_train, Y_train, "Train")

airline date_of_journey   source destination  dep_time arrival_time  \
0       Indigo      2019-05-09    Delhi      Cochin  06:50:00     16:10:00   
1    Air India      2019-04-03   Mumbai   Hyderabad  06:20:00     07:40:00   
2     Spicejet      2019-05-15  Kolkata    Banglore  11:15:00     18:30:00   
3  Jet Airways      2019-06-09    Delhi      Cochin  19:45:00     12:35:00   
4  Jet Airways      2019-06-18    Delhi      Cochin  09:50:00     13:05:00   

   duration  total_stops              additional_info   class_  price  
0       560          1.0                      No Info  Economy   6442  
1        80          0.0                      No Info  Economy   3100  
2       435          1.0                      No Info  Economy   8844  
3      1010          1.0  In-flight meal not included  Economy  10577  
4       195          0.0                      No Info  Economy   9564

In [217]:
export_data(X_test, Y_test, "Test")

airline date_of_journey   source destination  dep_time  \
0           Spicejet      2019-03-09  Chennai     Kolkata  08:20:00   
1        Jet Airways      2019-05-27    Delhi      Cochin  13:25:00   
2  Multiple Carriers      2019-05-09    Delhi      Cochin  11:30:00   
3             Indigo      2019-06-18  Kolkata    Banglore  21:25:00   
4  Multiple Carriers      2019-05-21    Delhi      Cochin  14:00:00   

  arrival_time  duration  total_stops additional_info   class_  price  
0     10:35:00       135          0.0         No Info  Economy   6300  
1     04:25:00       900          2.0         No Info  Economy  16704  
2     01:30:00       840          1.0         No Info  Economy  15078  
3     00:05:00       160          0.0         No Info  Economy   4804  
4     21:00:00       420          1.0         No Info  Economy  14067

In [218]:
export_data(X_val, Y_val, "Validation")

airline date_of_journey    source destination  dep_time arrival_time  \
0     Spicejet      2019-06-01  Banglore       Delhi  05:55:00     08:35:00   
1  Jet Airways      2019-06-21  Banglore       Delhi  17:45:00     20:45:00   
2  Jet Airways      2019-03-12  Banglore   New Delhi  05:45:00     20:20:00   
3  Jet Airways      2019-06-12    Mumbai   Hyderabad  19:35:00     21:05:00   
4  Jet Airways      2019-06-27     Delhi      Cochin  15:00:00     12:35:00   

   duration  total_stops              additional_info   class_  price  
0       160          0.0                      No Info  Economy   3625  
1       180          0.0  In-flight meal not included  Economy   7754  
2       875          1.0                      No Info  Economy  13817  
3        90          0.0                      No Info  Economy   5678  
4      1295          1.0  In-flight meal not included  Economy  10262